### PR TITLE
Kernel: Handle symlink path resolution when path ends with "/" 

### PR DESCRIPTION
**Pull request by shannonbooth**
Fixes an assert when `ls /proc/self/` is run.

A null StringView would be passed in as the substring to ascertain the remaining path to resolve.

### CODE REVIEW

File: Kernel/FileSystem/VirtualFileSystem.cpp [hunk 1]
```diff
@@ -720,12 +720,15 @@ KResultOr<NonnullRefPtr<Custody>> VFS::resolve_path(StringView path, Custody& ba
...
-            if (i + 1 == parts.size() - 1 && parts[i + 1].is_empty())
-                have_more_parts = false;
+            if (i == parts.size() - 1)
+                return symlink_target;
+
+            if (i == parts.size() - 2 && parts[i + 1].is_empty())
+                return symlink_target;
 
-            if (!have_more_parts)
+            if (parts[i + 1].is_null()) {
```
Comment (shannonbooth) [inline Kernel/FileSystem/VirtualFileSystem.cpp]: I'm not sure whether it's correct to just skip the part if it is null, or if it is just incorrect for it to be null in the first place...

**Comment by bugaevc** [comment]
As it was me who changed this code last (I think...), it would be nice of you @shannonbooth to cc me in this PR so that I wouldn't miss it. Please do that next time :)

**Comment by bugaevc** [comment]
TBH I'm not following this too: how is the new logic different (other than checking for `is_null()` instead of `is_empty()` — doesn't the former imply the latter?), and why wouldn't the existing code — namely `i + 1 == parts.size() - 1 && parts[i + 1].is_empty()` — handle `/proc/self/` properly?

**Comment by shannonbooth** [comment]
Yeah so the logic is no different that beforehand other than the addition of the null check. I also assumed that is_empty() also implied the is_null() check, but apparently it doesn't. Maybe that is simply the issue then? Honestly I'm not sure, I'm a little confused at how the resolve_path() function works.

**Comment by shannonbooth** [comment]
I changed around the existing logic because I didn't understand it at first until I realized all it meant was that we would return if we were:

- at the last part 
- or at the second to last part, and the next part was empty

Maybe that was a bad change and just made it more confusing. But yeah without the if part[i + 1].is_empty(), if you run `cat /proc/self/` AK/SubstringView.cpp will assert:

```
ASSERTION FAILED: !substring.is_null()
```

**Comment by shannonbooth** [comment]
I changed around the existing logic because I didn't understand it at first until I realized all it meant was that we would return if we were:

- at the last part 
- or at the second to last part, and the next part was empty

The reason why the existing code doesn't cover the assert being thrown is because the next part can be empty when we aren't at the second to last character. 

Now that I think about it I'm not sure if that behavior is what is broken, or if the lack of a check for it is. With these changes `ls /proc/self` seems to work though.

And I guess maybe it would make more sense if the check was changed to is_empty() instead of is_null()? I just kinda went with is_null() because I knew the assert was happening because it was null.

**Comment by shannonbooth** [comment]
Actually yeah my PR is definitely wrong, what's even the point in:
```
if (i == parts.size() - 2 && parts[i + 1].is_empty())
    return symlink_target;
```
With
```
if (parts[i + 1].is_empty())
    return sylink_target;
```
Right below it :nauseated_face:

**Comment by bugaevc** [comment]
> And I guess maybe it would make more sense if the check was changed to is_empty() instead of is_null()?

But isn't that exactly what the existing code does?

**Comment by shannonbooth** [comment]
Okay so here's a simplified version of what my changes are actually doing, which should make more sense. I'm not sure if it is actually correct though -- because why is the part null in the first place.

**Comment by bugaevc** [comment]
So this new diff looks like it's changing how `/foo//bar` is resolved, and... it shouldn't be changing that? Consecutive slashes should be treated just like a single slash.

> because why is the part null in the first place.

If you can reproduce getting null substrings from `StringView::split_view()`, that's definitely a bug there.

**Comment by shannonbooth** [comment]
What is happening is that a null substring is being passed into substring_view_starting_from_substring(const StringView& substring), which is causing the kernel to assert.

Just run `ls /proc/self/`, and you'll get `ASSERTION FAILED: remaining_characters >= m_characters` in that substring function, because the substring being passed in is null (which is why I added the assert in those functions to make it more clear what went wrong)

I'm open to ideas for a fix if my changes aren't the way forward.

**Comment by bugaevc** [comment]
> What is happening is that a null substring is being passed into `substring_view_starting_from_substring(const StringView& substring)`

The question is, where does that null substring come from?

> Just run `ls /proc/self/`

I'm not at my computer right now so I cannot investigate myself ;(

**Comment by bugaevc** [comment]
> If you can reproduce getting null substrings from `StringView::split_view()`, that's definitely a bug there.

There we have it: https://github.com/SerenityOS/serenity/blob/2de19342e7b419798674a64082027acc1ed8bae7/AK/StringView.cpp#L38-L39

This would be fine if we wouldn't want to pass these strings into `substring_view_starting_*_substring()`. We do, so let's pass a proper empty substring at the end of `this` here.

**Comment by shannonbooth** [comment]
> > If you can reproduce getting null substrings from `StringView::split_view()`, that's definitely a bug there.
> 
> There we have it:
> 
> https://github.com/SerenityOS/serenity/blob/2de19342e7b419798674a64082027acc1ed8bae7/AK/StringView.cpp#L38-L39
> 
> This would be fine if we wouldn't want to pass these strings into `substring_view_starting_*_substring()`. We do, so let's pass a proper empty substring at the end of `this` here.

Sorry I think I'm being a bit dense here, I'm not sure what you mean

**Comment by bugaevc** [comment]
There exist null strings (whose character pointer is `nullptr`) and _empty_ strings (whose length is zero). A null string always has to be empty (otherwise that'd imply that it's possible to read some bytes from `nullptr`), but an empty string doesn't have to be null (it can, or it can have a valid data pointer but zero length).

`String::empty()` creates a null string (I would assume), as does `{}`.

`substring_view_starting_*_substring()` methods, unlike basically every other piece of code, really care about *where* the substring you pass to them is (where its character pointer points to).

We want to be able to use the result of `split_view()` with `substring_view_starting_*_substring()`. This means it cannot just create a null substring when it really wants an empty substring at the end of the original string. So I'm saying we should return that instead of a null string.

**Comment by awesomekling** [comment]
> `String::empty()` creates a null string (I would assume), as does `{}`.

Actually `String::empty()` returns "the empty string", which is a singleton object representing "" (not null, zero length.) :)

**Comment by shannonbooth** [comment]
Closing because I don't know how to proceed :disappointed: refer to #970

**Pull request closed without merging**